scipy/linalg/_linsolve.pyx.in

```diff
@@ -0,0 +1,603 @@
+        free(cwork)
+        {{else}}
+        free(ipiv)
+        {{endif}}
+        raise LinAlgError('LAPACK routine {{typeprefix}}getrs reported an illegal value in {}-th argument.'.format(-info[0]))
+
+    {{if typeprefix in ['s', 'd']}}
+    # Recycle ipiv as iwork
+    {{typeprefix}}gecon(norm, n, a, n, &anorm, &rcond, work, ipiv, info);
+    free(ipiv)
+    {{else}}
+    {{typeprefix}}gecon(norm, n, a, n, &anorm, &rcond, cwork, work, info);
+    free(cwork)
+    {{endif}}
+    free(work)
+
+    if info[0] != 0:
+        raise LinAlgError('LAPACK routine {{typeprefix}}gecon reported an illegal value in {}-th argument.'.format(-info[0]))
+
+    # recycle anorm for eps
+    {{if typeprefix in ['s', 'c']}}
+    anorm = slamch('E')
+    {{else}}
+    anorm = dlamch('E')
+    {{endif}}
+    if rcond < anorm:  # rcond < eps
+        with gil:
+            warn('Ill-conditioned matrix (rcond={:.6g}): result may not be accurate.'.format(rcond), LinAlgWarning)
+
+    return 0
+
+
+{{endfor}}
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+cpdef inline int lusolve(lapack_t[:, ::1] a,
+                         lapack_t[::1, :]b,
+                         bint a_is_c) nogil except -1:
+    """
+    LU solution is a straightforward call to two LAPACK
+    routines first ?getrf and then ?getrs. Here to detect the
+    ill-conditioned problems we squeeze a ?gecon call after
+    getrf.
+
+    Since A is in row-major order, we factorize A.T and use
+    transposed version in getrs depending is a_is_c array or not.
+    """
+    cdef char* trans = 'T' if a_is_c else 'N'
+    cdef char* norm = '1' if a_is_c else 'I'
+    cdef int info, res
+    cdef int n = <int>a.shape[0]
+    cdef int nrhs = <int>b.shape[1]
+
+    if lapack_t is float:
+        res = lusolve_s(trans, norm, &n, &nrhs, &info, &a[0, 0], &b[0, 0])
+    elif lapack_t is double:
+        res = lusolve_d(trans, norm, &n, &nrhs, &info, &a[0, 0], &b[0, 0])
+    elif lapack_t is floatcomplex:
+        res = lusolve_c(trans, norm, &n, &nrhs, &info, &a[0, 0], &b[0, 0])
+    else:
+        res = lusolve_z(trans, norm, &n, &nrhs, &info, &a[0, 0], &b[0, 0])
+
+    return res
+
+
+# ########################
+# internal diag function
+# ########################
+{{for typename, typeprefix, typenamescdz in zip(typenames, prefixes, typenames_scdz)}}
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+def diagsolve_{{typeprefix}}(({{typename}})[::1] a, ({{typename}})[::1, :]b):
+    cdef int r, c, n = <int>len(a), m = <int>b.shape[1]
+    cdef {{typenamescdz}}[::1] diags_abs
+    cdef {{typenamescdz}} diags_min, rcond
+
+    for r in xrange(n):
+        if a[r] == 0.:
+            raise LinAlgError("Matrix is singular")
+
+    with nogil:
+        for r in xrange(n):
+            for c in xrange(m):
+                # Cython doesn't like complex numbers and prefix ops !?
+                b[r, c] = b[r, c] / a[r]
```

Comment:
Could use tempita to fall back on this for complex numbers and take advantage of inplace operators for real-valued arrays.  But a good compiler might already be doing this anyways

Reply:
Yup, cython does handle += etc for complex values. Apparently there is a reason, but the explanation I once got from cython-user mailing list was... Ok, there was a reason, that's all I understood